xunit.runner.wpf/ViewModel/MainViewModel.cs

```diff
@@ -400,85 +302,108 @@ private bool IsConfigFile(string fileName)
         private bool CanExecuteRun()
             => !IsBusy && TestCases.Any();
 
-        private async void OnExecuteRun()
+        private void OnExecuteRun()
         {
             try
             {
-                IsBusy = true;
                 TestsCompleted = 0;
                 TestsPassed = 0;
                 TestsFailed = 0;
                 TestsSkipped = 0;
                 CurrentRunState = TestState.NotRun;
+                RunTests();
                 Output = string.Empty;
-                await Task.Run(() => RunTestsInBackground());
             }
             catch (Exception ex)
             {
                 MessageBox.Show(ex.ToString());
             }
-            finally
-            {
-                IsBusy = false;
-                IsCancelRequested = false;
-            }
         }
 
-        private void RunTestsInBackground()
+        private void RunTests()
         {
+            Debug.Assert(this.cancellationTokenSource == null);
+            Debug.Assert(this.testSessionList.Count == 0);
+
             foreach (var tc in TestCases)
             {
                 tc.State = TestState.NotRun;
             }
 
-            var selectedAssemblies = TestCases.ToLookup(tc => tc.AssemblyFileName);
-            using (AssemblyHelper.SubscribeResolve())
+            // TODO: Need a way to filter based on traits, selected test cases, etc ...  For now we just run
+            // everything. 
```

Comment:
I have filtered `TestCases` working on my dev box.  I thought I pushed it before I left today but apparently not.  Overall though it establishes the pattern of sending a good deal of information between UI and worker process.  Adding traits will be a small amount of work on top of that.